src/vs/editor/test/browser/view/minimapFontCreator.ts

```diff
@@ -4,31 +4,20 @@
  *--------------------------------------------------------------------------------------------*/
 
 import { RGBA8 } from 'vs/editor/common/core/rgba';
-import { Constants, MinimapCharRenderer } from 'vs/editor/common/view/minimapCharRenderer';
-import { getOrCreateMinimapCharRenderer } from 'vs/editor/common/view/runtimeMinimapCharRenderer';
-import { MinimapCharRendererFactory } from 'vs/editor/test/common/view/minimapCharRendererFactory';
+import { Constants, MinimapCharRenderer } from 'vs/editor/browser/viewParts/minimap/minimapCharRenderer';
+import { MinimapCharRendererFactory } from 'vs/editor/browser/viewParts/minimap/minimapCharRendererFactory';
 
-let canvas = <HTMLCanvasElement>document.getElementById('my-canvas');
-let ctx = canvas.getContext('2d')!;
-
-canvas.style.height = 100 + 'px';
-canvas.height = 100;
-
-canvas.width = Constants.CHAR_COUNT * Constants.SAMPLED_CHAR_WIDTH;
-canvas.style.width = (Constants.CHAR_COUNT * Constants.SAMPLED_CHAR_WIDTH) + 'px';
-
-ctx.fillStyle = '#ffffff';
-ctx.font = 'bold 16px monospace';
-for (let chCode = Constants.START_CH_CODE; chCode <= Constants.END_CH_CODE; chCode++) {
-	ctx.fillText(String.fromCharCode(chCode), (chCode - Constants.START_CH_CODE) * Constants.SAMPLED_CHAR_WIDTH, Constants.SAMPLED_CHAR_HEIGHT);
-}
-
-let sampleData = ctx.getImageData(0, 4, Constants.SAMPLED_CHAR_WIDTH * Constants.CHAR_COUNT, Constants.SAMPLED_CHAR_HEIGHT);
-let minimapCharRenderer = MinimapCharRendererFactory.create(sampleData.data);
+let sampleData = MinimapCharRendererFactory.createSampleData();
+let minimapCharRenderer1x = MinimapCharRendererFactory.createFromSampleData(sampleData.data, 1);
+let minimapCharRenderer2x = MinimapCharRendererFactory.createFromSampleData(sampleData.data, 2);
+let minimapCharRenderer4x = MinimapCharRendererFactory.createFromSampleData(sampleData.data, 4);
+let minimapCharRenderer6x = MinimapCharRendererFactory.createFromSampleData(sampleData.data, 6);
 
 renderImageData(sampleData, 10, 100);
-renderMinimapCharRenderer(minimapCharRenderer, 400);
-renderMinimapCharRenderer(getOrCreateMinimapCharRenderer(), 600);
+renderMinimapCharRenderer(minimapCharRenderer1x, 400, 1);
+renderMinimapCharRenderer(minimapCharRenderer2x, 500, 2);
+renderMinimapCharRenderer(minimapCharRenderer4x, 600, 4);
+renderMinimapCharRenderer(minimapCharRenderer6x, 750, 8);
```

Comment:
> > Generating fonts at runtime does incur a small performance penalty, taking
about 0.6m at 1x scale and 0.9ms at 4x scale on my machine to create the font
the first time we render a minimap.

Shouldn't we only be rendering the scale we need and reacting when the scale setting or device pixel ratio changes?

Reply:
Additionally, this can vary wildly based on hardware and Chromium version based on my experience.

Reply:
> Shouldn't we only be rendering the scale we need and reacting when the scale setting or device pixel ratio changes?

This is what we do! This file is a 'demo' page, previously used to generate the hardcoded image maps, and I put several more scales here now to assist with development. In the editor, the minimap calls `MinimapCharRendererFactory.create(scale)` when it resolves its rendering setts. The factory caches a singleton of the char renderer, so that it only creates one as needed (instead of rerendering the font for every minimap instance)